packages/amplify-e2e-core/src/utils/nexpect.ts

```diff
@@ -357,12 +367,17 @@ function chain(context: Context): ExecutionContext {
     delay(milliseconds: number): ExecutionContext {
       const _delay: ExecutionStep = {
         fn: () => {
-          const startCallback = Date.now();
-
-          while (Date.now() - startCallback < milliseconds) {
-            // empty
-          }
-
+          /*
+            Code below is workaround for lack of synchronous sleep() in JS.
+            It has nothing to do with atomicity nor with buffers. It's just using these
+            built-in APIs to wait synchronously. I.e. it waits number of milliseconds for
+            byte transition in the buffer that never happens.
+            This replaced active spin-lock that was here before
+            and was burning 100% CPU while waiting that led to spawned CLI starvation.
+            If this way of delaying becomes insufficient then this module should
+            be refactored and implement fully asynchronous input handlers.
+          */
+          Atomics.wait(new Int32Array(new SharedArrayBuffer(4)), 0, 0, milliseconds);
```

Comment:
If you wonder how did I find this https://github.com/sindresorhus/sleep-synchronously/blob/f9cac30d9d4b2019ee766a9e3d904b6b96010531/index.js#L2 .

Reply:
TIL